contracts/AbstractRegistrar.sol

```diff
@@ -0,0 +1,97 @@
+pragma solidity ^0.4.24;
+
+import "@ensdomains/ENS/contracts/ENS.sol";
+import "@ensdomains/dnssec-oracle/contracts/DNSSEC.sol";
+import "@ensdomains/dnssec-oracle/contracts/BytesUtils.sol";
+import "@ensdomains/dnsregistrar/contracts/DNSClaimChecker.sol";
+
+contract AbstractRegistrar {
+
+    using BytesUtils for bytes;
+
+    struct Record {
+        address submitter;
+        address addr;
+        bytes32 proof;
+        bytes32 name;
+        bytes32 label;
+        bytes32 node;
+        uint256 submitted;
+    }
+
+    uint16 constant CLASS_INET = 1;
+    uint16 constant TYPE_TXT = 16;
+
+    ENS public ens;
+    DNSSEC public oracle;
+
+    uint256 public cooldown;
+    uint256 public stake;
+
+    /// label => record
+    mapping (bytes32 => Record) public records;
+
+    event Submitted(bytes32 indexed node, address indexed owner, bytes proof, bytes dnsname);
+    event Claim(bytes32 indexed node, address indexed owner);
+    event Challenged(bytes32 indexed node, address indexed challenger);
+
+    constructor(ENS _ens, DNSSEC _dnssec, uint256 _cooldown, uint256 _stake) public {
+        ens = _ens;
+        oracle = _dnssec;
+        cooldown = _cooldown;
+        stake = _stake;
+    }
+
+    /// @notice This function allows the user to submit a DNSSEC proof for a certain amount of ETH.
+    function _submit(bytes name, bytes proof, address addr) internal {
+        bytes32 label;
+        bytes32 node;
+        (label, node) = DNSClaimChecker.getLabels(name);
+
+        records[keccak256(node, label)] = Record({
+            submitter: msg.sender,
+            addr: addr,
+            proof: keccak256(proof),
+            name: keccak256(name),
+            label: label,
```

Comment:
do we really want to do any form of computations in an optimistic contracts submission phase?

Reply:
You're already doing the same computation in this function - if you do it in the submission phase, you're using less total gas, because you don't need to store the values.

Reply:
@Arachnid so are you saying we should calculate the labels in the commit function?